integration_tests/src/main/scala/com/nvidia/spark/rapids/tests/common/BenchUtils.scala

```diff
@@ -0,0 +1,176 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.nvidia.spark.rapids.tests.common
+
+import java.io.{File, FileOutputStream}
+import java.time.Instant
+import java.util.concurrent.TimeUnit.NANOSECONDS
+
+import scala.collection.mutable.ListBuffer
+
+import org.json4s.DefaultFormats
+import org.json4s.jackson.JsonMethods.parse
+import org.json4s.jackson.Serialization.writePretty
+
+import org.apache.spark.{SPARK_BUILD_USER, SPARK_VERSION}
+import org.apache.spark.sql.{DataFrame, Row, SparkSession}
+
+object BenchUtils {
+
+  /**
+   * Run the specified number of cold and hot runs and record the timings and summary of the
+   * query and results to file, including all Spark configuration options and environment
+   * variables.
+   */
+  def runBench(
+      spark: SparkSession,
+      queryRunner: SparkSession => DataFrame,
+      queryDescription: String,
+      filenameStub: String,
+      numColdRuns: Int = 1,
+      numHotRuns: Int = 3,
+      maxResultsToRecord: Int = 100): Unit = {
+
+    val now = Instant.now()
+
+    var df: DataFrame = null
+    var results: Seq[Row] = null
+    val coldRunElapsed = new ListBuffer[Long]()
+    for (i <- 0 until numColdRuns) {
+      println(s"*** Start cold run $i:")
+      val start = System.nanoTime()
+      df = queryRunner(spark)
+      results = df.collect
+      val end = System.nanoTime()
+      val elapsed = NANOSECONDS.toMillis(end - start)
+      coldRunElapsed.append(elapsed)
+      println(s"*** Cold run $i took $elapsed msec.")
+    }
+
+    val hotRunElapsed = new ListBuffer[Long]()
+    for (i <- 0 until numHotRuns) {
+      println(s"*** Start hot run $i:")
+      val start = System.nanoTime()
+      df = queryRunner(spark)
+      results = df.collect
+      val end = System.nanoTime()
+      val elapsed = NANOSECONDS.toMillis(end - start)
+      hotRunElapsed.append(elapsed)
+      println(s"*** Hot run $i took $elapsed msec.")
+    }
+
+    for (i <- 0 until numColdRuns) {
+      println(s"Cold run $i took ${coldRunElapsed(i)} msec.")
+    }
+    println(s"Average cold run took ${coldRunElapsed.sum.toDouble/numColdRuns} msec.")
+
+    for (i <- 0 until numHotRuns) {
+      println(s"Hot run $i took ${hotRunElapsed(i)} msec.")
+    }
+    println(s"Average hot run took ${hotRunElapsed.sum.toDouble/numHotRuns} msec.")
+
+    // write results to file
+    val filename = s"$filenameStub-${now.toEpochMilli}.json"
+    println(s"Saving results to $filename")
+
+    // try not to leak secrets
+    val redacted = Seq("TOKEN", "SECRET", "PASSWORD")
+    val envVars: Map[String, String] = sys.env
+        .filterNot(entry => redacted.exists(entry._1.toUpperCase.contains))
```

Comment:
I'd almost want to see: "default variables for spark at SHA X" and "this is what we did for our benchmark that wasn't in the default variables". This can totally by follow on work.

Reply:
I think this is something we could do when we generate reports from the raw data